pkg/component/runtime/manager_test.go

```diff
@@ -1162,14 +1175,312 @@ LOOP:
+			{
+				ID:       "fake-input",
+				Type:     client.UnitTypeInput,
+				LogLevel: client.UnitLogLevelTrace,
+				Config: component.MustExpectedConfig(map[string]interface{}{
+					"type":    "fake",
+					"state":   int(client.UnitStateHealthy),
+					"message": "Fake Healthy",
+				}),
+			},
+		},
+	}
+
+	subCtx, subCancel := context.WithCancel(context.Background())
+	defer subCancel()
+	subErrCh := make(chan error)
+	go func() {
+		killed := false
+
+		sub := m.Subscribe(subCtx, "fake-default")
+		for {
+			select {
+			case <-subCtx.Done():
+				return
+			case state := <-sub.Ch():
+				t.Logf("component state changed: %+v", state)
+				if state.State == client.UnitStateFailed {
+					if !killed {
+						subErrCh <- fmt.Errorf("component failed: %s", state.Message)
+					}
+				} else {
+					unit, ok := state.Units[ComponentUnitKey{UnitType: client.UnitTypeInput, UnitID: "fake-input"}]
+					if ok {
+						if unit.State == client.UnitStateFailed {
+							if !killed {
+								subErrCh <- fmt.Errorf("unit failed: %s", unit.Message)
+							}
+						} else if unit.State == client.UnitStateHealthy {
+							// force the input to exit and it should be restarted
+							if !killed {
+								killed = true
+
+								r := regexp.MustCompile(`pid \'(?P<pid>\d+)\'`)
+								rp := r.FindStringSubmatch(state.Message)
+								t.Logf("triggering kill through config on pid %s", rp)
+								comp.Units[0].Config = component.MustExpectedConfig(map[string]interface{}{
+									"type":    "fake",
+									"state":   int(client.UnitStateHealthy),
+									"message": "Fake Healthy",
+									"kill":    rp[1],
+								})
+								err := m.Update([]component.Component{comp})
+								if err != nil {
+									subErrCh <- err
+								}
+							} else {
+								// got back to healthy after kill
+								subErrCh <- nil
+							}
+						} else if unit.State == client.UnitStateStarting || unit.State == client.UnitStateStopped {
+							// acceptable
+						} else {
+							// unknown state that should not have occurred
+							subErrCh <- fmt.Errorf("unit reported unexpected state: %v", unit.State)
+						}
+					} else {
+						subErrCh <- errors.New("unit missing: fake-input")
+					}
+				}
+			}
+		}
+	}()
+
+	defer drainErrChan(errCh)
+	defer drainErrChan(subErrCh)
+
+	startTimer := time.NewTimer(100 * time.Millisecond)
+	defer startTimer.Stop()
+	select {
+	case <-startTimer.C:
+		err = m.Update([]component.Component{comp})
+		require.NoError(t, err)
+	case err := <-errCh:
+		t.Fatalf("failed early: %s", err)
+	}
+
+	endTimer := time.NewTimer(1 * time.Minute)
+	defer endTimer.Stop()
+LOOP:
+	for {
+		select {
+		case <-endTimer.C:
+			t.Fatalf("timed out after 30 seconds")
+		case err := <-errCh:
+			require.NoError(t, err)
+		case err := <-subErrCh:
+			require.NoError(t, err)
+			break LOOP
+		}
+	}
+
+	subCancel()
+	cancel()
+
+	err = <-errCh
+	require.NoError(t, err)
+}
+
+func TestManager_FakeInput_KeepsRestarting(t *testing.T) {
```

Comment:
Did this or the other test ever reproduce the problem? What about if you reverted my `ClearPendingCheckin` change that didn't completely fix the problem? Does it catch it then or not at all?

Reply:
Do the tests catch the issue with no fixes applied at all? (is what I was trying to ask above)

Reply:
I just tried it with your change removed and sadly these did not show the symptoms either. I wanted to keep the tests in the change because overall I thought they still exercised the code path well.

Reply:
Agreed I like them too, maybe if we ran them for long enough they'd catch it.